src/main/java/com/github/f0xdx/Schemas.java

```diff
@@ -113,4 +119,101 @@ public static SchemaBuilder toBuilder(@NonNull Schema schema) {
 
     return builder;
   }
+
+  /**
+   * Returns a schema for storing the message headers
+   *
+   * @param headers
+   * @return
+   */
+  public static Schema forHeaders(Headers headers) {
+    final Map<String, List<Schema>> schemas = new HashMap<>();
+    StreamSupport.stream(headers.spliterator(), false)
+        .forEach(
+            header -> {
+              if (!schemas.containsKey(header.key())) {
+                schemas.put(header.key(), new ArrayList<>());
+              }
+              schemas.get(header.key()).add(header.schema());
+            });
+    final SchemaBuilder builder = SchemaBuilder.struct();
+    schemas.forEach((key, value) -> builder.field(key, mergeSchemas(value)));
+    return builder.build();
+  }
+
+  private static Schema mergeSchemas(List<Schema> headers) {
+    Schema schema = null;
+    for (Schema headerSchema : headers) {
+      schema = mergeSchemas(schema, headerSchema);
+    }
+    return SchemaBuilder.array(schema).optional().build();
+  }
+
+  private static Schema mergeSchemas(Schema schema, Schema addition) {
+    if (schema == null) {
+      return addition;
+    }
+    switch (addition.type()) {
+      case FLOAT32:
+      case FLOAT64:
+        if (schema.type().equals(Schema.Type.FLOAT32)
+            || schema.type().equals(Schema.Type.FLOAT64)) {
+          return Schema.OPTIONAL_FLOAT64_SCHEMA;
+        }
+      case INT8:
+      case INT16:
+      case INT32:
+      case INT64:
+        if (Arrays.asList(Schema.Type.INT8, Schema.Type.INT16, Schema.Type.INT32, Schema.Type.INT64)
+            .contains(schema.type())) {
+          return Schema.OPTIONAL_INT64_SCHEMA;
+        }
+      case STRUCT:
+        if (schema.type() == Schema.Type.STRUCT) {
+          return mergeStructs(schema, addition);
+        }
+      case MAP:
+        if (schema.type() == Schema.Type.MAP) {
+          return SchemaBuilder.map(
+                  mergeSchemas(schema.keySchema(), addition.keySchema()),
+                  mergeSchemas(schema.valueSchema(), addition.valueSchema()))
+              .optional()
+              .build();
+        }
+      case ARRAY:
+        if (schema.type() == Schema.Type.MAP) {
```

Comment:
Guess this should be

```suggestion
        if (schema.type() == Schema.Type.ARRAY) {
```